helper/logging/logging_http_transport.go

```diff
@@ -0,0 +1,210 @@
+// The logging is done via `tflog`, that is part of the terraform-plugin-log
+// library, included by this SDK.
+//
+// The request/response is logged via tflog.SubsystemDebug, and the `subsystem`
+// is the one provide here.
+//
+// IMPORTANT: For logging to work, it's mandatory that each http.Request it handles
+// is configured with the Context (i.e. `Request.WithContext(ctx)`)
+// that the SDK passes into all resources/data-sources/provider entry-points
+// (i.e. schema.Resource fields like `CreateContext`, `ReadContext`, etc.).
+func NewLoggingHTTPTransport(subsystem string, t http.RoundTripper) *loggingHTTPTransport {
+	return &loggingHTTPTransport{subsystem, t, nil}
+}
+
+// WithConfigureRequestContext allows to optionally configure a callback ConfigureReqCtxFunc.
+// This is used by the underlying structure to allow user to configure the
+// http.Request context.Context, before the request is executed and the logger
+// tflog.SubsystemDebug invoked.
+//
+// Log entries will be structured to contain:
+//
+//   * the HTTP message first line, broken up into "fields" (see `FieldHttp*` constants)
+//   * the Headers, each as "fields" of the log
+//   * the Request/Response Body as "message" of the log
+//
+// For example, here is how to add an extra field to each log emitted here,
+// as well as masking fields that have a specific key:
+//
+//   t.WithConfigureRequestContext(func (ctx context.Context, subsystem string) context.Context {
+//     ctx = tflog.SetField(ctx, "additional_key", "additional_value")
+//     ctx = tflog.MaskFieldValuesWithFieldKeys(ctx, "secret", "token", "Authorization")
+//   })
+//
+func (t *loggingHTTPTransport) WithConfigureRequestContext(callback ConfigureReqCtxFunc) *loggingHTTPTransport {
+	t.configureReqCtx = callback
+	return t
+}
+
+func decomposeRequestBytes(b []byte) (string, map[string]interface{}) {
+	parts := strings.Split(string(b), "\r\n")
+
+	// We will end up with a number of fields equivalent to the number of parts + 1:
+	// - the first will be split into 3 parts
+	// - the last 2 are "end of headers" and "body"
+	// - one extra for the operation type
+	fields := make(map[string]interface{}, len(parts)+1)
+	fields[FieldHttpOperationType] = OperationHttpRequest
+
+	// HTTP Request Line
+	reqLineParts := strings.Split(parts[0], " ")
+	fields[FieldHttpRequestMethod] = reqLineParts[0]
+	fields[FieldHttpRequestUri] = reqLineParts[1]
+	fields[FieldHttpRequestVersion] = reqLineParts[2]
+
+	// HTTP Request Headers
+	var i int
+	for i = 1; i < len(parts)-2; i++ {
+		// Check if we reached the end of the headers
+		if parts[i] == "" {
+			break
+		}
+
+		headerParts := strings.Split(parts[i], ": ")
+		fields[headerParts[0]] = headerParts[1]
+	}
+
+	// HTTP Response Body: the last part is always the body (can be empty)
+	return parts[len(parts)-1], fields
+}
+
+func decomposeResponseBytes(b []byte) (string, map[string]interface{}) {
+	parts := strings.Split(string(b), "\r\n")
```

Comment:
Will look into it

Reply:
Actually, `ReadMIMEHeader()` doesn't like the full body of a req/res (what I have here), but only the sub-section that contains actual headers.

So I can't use it to read the `HTTP Request Line`, nor the `HTTP Message Status Line`.
I'll see if it makes sense to ue it for the headers though.

Reply:
I was peeking at the `ReadResponse` function in https://go.dev/src/net/http/response.go, which looks like it does `ReadLine()` then `ReadMIMEHeaders()`, but then I realized you already have a [`http.Response` object](https://pkg.go.dev/net/http#Response). Can we just use that information? How about with the request?

Reply:
The usage of `httputil.Dump*` functionality in the original implementation may have been to just simplify the plaintext-only representation there. Since we want to provide richer information, I think we can use the higher-level abstractions available in `http.Request` and `http.Response`. 😄

Reply:
I haven't finished yet, but I think you are right: for `http.Response` shouldn't be too tricky to just use the struct as is.

For `http.Request` I don't know yet, but will give it another go.

The key issue though seems to be the bit that those `Dump*` function did with the body: being the body a `ReadCloser`, it's tricky to grab that without interfering with the request/response. Those `Dump*` utilities do lots of "tricks" around the body to ensure they return a copy, without affecting the underlying buffer and the expectation of http client that is sending/receiving.

I'll have to do the same to avoid disrupting things, and this is the main reason I was sticking with the `Dump*` methods: splitting strings around the most well-formed standard heading in the industry, felt a safe long term bet.

But 🙆‍♂️ .

Reply:
Update (focusing on the response right now, but request isn't much different).

1. A detail that we missed so far is that `Response.Body` contains the WHOLE body of the response. Meaning, while I can use `Response.Header`, `Response.Status` etc. to populate most fields, `Response.Body` contains the whole body of the request (html/json/whatever), plus all the same headers above it.

2. In addition to that, the process to read the bytes out of a `Response.Body`, while preserving the response to that the client can consume it, it's requiring me to re-implement `httputil.DumpResponse` entirely. So I'm planning to keep using it instead, to grab the `Body`, while using the `Response.Header` (etc) fields to populate the other fields.

What do you advice to do about 1. and what do you think of 2.?